src/gui/widget/TextField.hpp

```diff
@@ -0,0 +1,120 @@
+#ifndef PUTKARTS_GUI_Widget_TextField_HPP
+#define PUTKARTS_GUI_Widget_TextField_HPP
+
+#include <string>
+
+#include "gui/widget/Widget.hpp"
+
+#include "SFML/Graphics.hpp"
+
+namespace GUI {
+	namespace Widget {
+		class TextField;
+	}
+}
+
+/**
+ * One row textfield.
+ */
+class GUI::Widget::TextField: public Widget {
+	/** The text */
+	sf::Text text;
+
+	/** Cursor index */
+	unsigned int cursorIndex;
+
+	/**
+	 * Get size of font.
+	 *
+	 * @return Font size
+	 */
+	unsigned int fontSize() const;
+
+	/**
+	 * Get width of borders.
+	 *
+	 * @return Border size
+	 */
+	unsigned int borderSize() const;
+
+	/**
+	 * Get character size.
+	 *
+	 * @return Character size
+	 */
+	sf::Vector2f characterSize() const;
+
+	/**
+	 * Get start position of text on rendering.
+	 *
+	 * @return Text start position
+	 */
+	sf::Vector2f textStartPos() const;
+
+	/**
+	 * Get maximum text width text can have on rendering.
+	 *
+	 * @return Maximum text width
+	 */
+	float maxTextWidth() const;
+
+	/**
+	 * Get text width on rendering.
+	 *
+	 * @param str Text
+	 * @return Text width
+	 */
+	float textWidth(const sf::String& str) const;
+
+	/**
+	 * Get text index from X coordinate.
+	 *
+	 * @param xPosition X coordinate
+	 * @return Text index
+	 */
+	unsigned int getIndex(const float& xPosition) const;
+public:
+
+	/**
+	 * Create Text field.
+	 *
+	 * @param str Text to set
+	 * @param x X coordinate
+	 * @param y Y coordinate
+	 * @param width Width
+	 * @param height Height
+	 */
+	TextField(const std::string& str, float x, float y, float width, float height);
```

Comment:
You will have to add a CallbackType like in other widgets. Either report the changed string or just pass a reference to the widget. I'd prefer reporting the changed string, in which case you might want to modify the code so that the whole string doesn't need to be converted to/from UTF-8 every time there is a small change.